main/webservices/api/test_api.php

```diff
@@ -2,116 +2,531 @@
+            'First session'.time(),
+            '2019-01-01 00:00', '2019-08-31 00:00',
+            '2019-01-01 00:00', '2019-08-31 00:00',
+            '2019-01-01 00:00', '2019-08-31 00:00',
+            null, null
+        );
+        $secondSessionId = SessionManager::create_session(
+            'Second session'.time(),
+            '2019-09-01 00:00', '2019-12-31 00:00',
+            '2019-09-01 00:00', '2019-12-31 00:00',
+            '2019-09-01 00:00', '2019-12-31 00:00',
+            null, null
+        );
+
+        // assign unique distinct value in first field to each session
+        SessionManager::update_session_extra_field_value($firstSessionId, $FIRST_EXTRA_FIELD_NAME, $firstSessionId);
+        SessionManager::update_session_extra_field_value($secondSessionId, $FIRST_EXTRA_FIELD_NAME, $secondSessionId);
+
+        // assign the same value in second field to all sessions
+        $COMMON_VALUE = 'common value';
+        SessionManager::update_session_extra_field_value($firstSessionId, $SECOND_EXTRA_FIELD_NAME, $COMMON_VALUE);
+        SessionManager::update_session_extra_field_value($secondSessionId, $SECOND_EXTRA_FIELD_NAME, $COMMON_VALUE);
+
+        // assert that the correct session id is returned using each unique value
+        $this->assertSame($firstSessionId,
+            $this->integer($ACTION, [
+                'field_name' => $FIRST_EXTRA_FIELD_NAME,
+                'field_value' => $firstSessionId,
+            ]));
+        $this->assertSame($secondSessionId,
+            $this->integer($ACTION, [
+                'field_name' => $FIRST_EXTRA_FIELD_NAME,
+                'field_value' => $secondSessionId,
+            ]));
+
+        // assert search for common value in second field generates the right error message
+        $this->assertSame('MoreThanOneSessionMatched',
+            $this->errorMessageString($ACTION, [
+                'field_name' => $SECOND_EXTRA_FIELD_NAME,
+                'field_value' => $COMMON_VALUE,
+            ]));
+
+        // assert search for unknown value generates the right error message
+        $this->assertSame('NoSessionMatched', $this->errorMessageString($ACTION, [
+            'field_name' => $SECOND_EXTRA_FIELD_NAME,
+            'field_value' => 'non-existent value',
+        ]));
+
+        // clean up
+        SessionManager::delete($firstSessionId);
+        SessionManager::delete($secondSessionId);
+        $extraFieldModel->delete($firstExtraFieldId);
+        $extraFieldModel->delete($secondExtraFieldId);
+    }
+
+    /**
+     * @throws Exception when it cannot delete the test user
+     */
+    public function testUpdateUserFromUsername() {
+        $ACTION = 'update_user_from_username';
+        // create a user with initial data and extra field values
+        $LOGIN_NAME = 'testUser'.time();
+        $userId = UserManager::create_user('Initial first name', 'Initial last name', 5, 'initial.email@local', $LOGIN_NAME, 'xXxxXxxXX');
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Line exceeds maximum limit of 120 characters; contains 138 characters](https://app.codacy.com/app/chamilo/chamilo-lms/pullRequest?prid=4554450)